File: 프로그래머스/2/60057. 문자열 압축/문자열 압축.py
def solution(s):
    answer = 1000
    # n등분 드가자.
    # .count()써볼까
    length = len(s)
    
    if length == 1:
        return 1
    
    for num in range(1, len(s)):
        total = length % num
        time = length // num
        idx = 0
        cnt = 1
        word = ""
        while idx < num * time:
            if idx == 0:
                word = s[idx : idx + num]
                idx += num
                continue
            
            if s[idx : idx + num] == word:
                cnt += 1
                
            else:
                if cnt > 1:
                    total += (len(str(cnt)) + num)
                else:
                    total += num
                    
                cnt = 1
                word = s[idx : idx + num]
                
            idx += num
        
        if cnt > 1:
            total += (len(str(cnt)) + num)
        else:
            total += num
            
        answer = min(answer, total)

    return answer

File: 프로그래머스/2/60057. 문자열 압축/문자열 압축.java
import java.util.*;

class Solution {
    public int solution(String s) {
        int answer = 1000;
        int maxLen = s.length();
        if (s.length() == 1) return 1;
        
        for (int len = 1; len < maxLen; len++) {
            int cnt = 1;
            int idx = 0;
            String result = "";
            String curWord = s.substring(idx, idx + len);
            idx += len;
            
            while (idx <= maxLen - len) {
                String nextWord = s.substring(idx, idx + len);
                // System.out.println(curWord + " " + nextWord);
                if (curWord.equals(nextWord)) {
                    cnt++;
                } else {
                    // 두자리 수 나오면 어떡할건데
                    // 그냥 더해서 마지막에 length로 뽑자
                    if (cnt > 1) result += (cnt + curWord);
                    else result += curWord;
                    curWord = nextWord;
                    cnt = 1;
                }
                idx += len;
            }
            
            if (cnt > 1) result += (cnt + curWord);
            else result += curWord;
            
            if (idx < maxLen) {
                result += s.substring(idx);
            }
            
            // System.out.println("last     " + result);
            answer = Math.min(answer, result.length());
        }
        
        return answer;
    }
}